ethereum/solidity/contracts/lib/tree/binary/BinaryMerkleProof.sol

```diff
@@ -0,0 +1,12 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.0;
+
+/// @notice Merkle Tree Proof structure.
+struct BinaryMerkleProof {
+    // List of side nodes to verify and calculate tree.
+    bytes32[] sideNodes;
+    // The key of the leaf to verify.
+    uint256 key;
+    // The number of leaves in the tree
+    uint256 numLeaves;
```

Comment:
Don't the number of sideNodes give away the depth of the tree and hence the numLeaves?

Reply:
Oh I guess they do, yes.

Reply:
Oh wait, I thought you meant the length of the proof. No, the number of sidenodes don't tell you the depth of the tree unless the tree is perfect, which it may not be.

Reply:
Shouldn't the depth of the tree be provided by the validators, not the prover? https://bitslog.com/2018/06/09/leaf-node-weakness-in-bitcoin-merkle-tree-design/